src/Autofac/Core/Resolving/Pipeline/ResolvePipelineBuilder.cs

```diff
@@ -271,10 +270,10 @@ Action<ResolveRequestContextBase> Chain(Action<ResolveRequestContextBase> next,
 
                 return (ctxt) =>
                 {
-                    // Optimise the path depending on whether a tracer is attached.
-                    if (ctxt.TracingEnabled)
+                    // Optimise the path depending on whether diagnostics are enabled.
+                    if (ctxt.DiagnosticSource.MiddlewareDiagnosticsEnabled())
```

Comment:
This is tough because the thing being checked is looking to see if any subscribers have this enabled, right? And there could be multiple subscribers, which could be added at any time.

I could simplify the checks - check for literally the exact event being subscribed instead of one of a logical set. Right now I check for start or success or fail to determine if any subscribers exist, but in the majority case I'm guessing that's three "false" checks every time. I could switch it to check only for the specific event being issued. It'll change the flow a bit since you could subscribe to a failure event without subscribing to start or success, but maybe that's okay.

Reply:
I think it's probably justified for performance reasons to say "If you change diagnostic subscriptions to middleware events, that change  may not take effect until the next resolve operation.

Especially since the 99% use case is going to be turning it on once, for all events, and never changing it.

Reply:
So evaluate all the subscriptions at the operation level and pass that all down through the stack? I think that might add some complexity. Let me see about making the checks more precise and see where that leaves us. I also have a not-totally-formed idea about using flags internally to do bitwise checks and speed up comparison.

Reply:
Honestly I was just thinking that if someone changes the set of suscribers or subscribed events such that our checking method returns a different value, we...might not raise some middleware events during that transition. But then if someone did change that set during a request, they would have missed related operation and request messages anyway, which could introduce different problems.

Is this sort of thing what the DiagnosticSource `Activity` functionality is for?

Fundamentally it's whatever gets us the perf saving with no subscribers, so whatever approach works.

Are we anticipating encouraging the use of diagnostics in production? If so, do we need to bench the diagnostic source capture?

Reply:
I thought about using Activity but it didn't seem worth the allocations.

I'm less concerned about changing subscriptions in the middle of the request than I am about trying to be too smart in the caching mechanism such that we stop responding correctly to observables or it becomes hard to add a new event in here. I could see wanting diagnostic events for all the regular lifetime events, too.

I don't think diagnostics in prod would be recommended. More like "turn this on if you find a problem and turn it back off" or something.

Reply:
Realized now that I was basically checking enabled _twice_ for any given write.

First time in `if(source.MiddlewareDiagnosticsEnabled())` and later in the actual call to write the event. Same for operation and request.

I've removed the aggregate checks and just leave it to the call to write the event. This should remove some of the overhead. **Let's see where this puts us perf-wise and I can try to tweak some more if it looks like it still needs help.**

I'm trying to not cart too much additional context around if possible, and while a single Boolean isn't that big a deal, the orchestration of keeping the cache up to date with the actual settings seems like it could be a challenge long-term, similar to how container rebuilding got to be a bit sticky and unpredictable.

I can see where there is a use case for success or failure instrumentation, like perf counters or something, to track stats. That'd mean subscribing to just success/failure events but maybe not start events. The ability to have that granularity seems interesting to me, rather than just "all on" or "all off" for a whole request.

I can also see that we may want to support (at some point) the [ability to conditionally determine if you're 'enabled' based on the data itself](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.diagnosticlistener.isenabled?view=netcore-3.1#System_Diagnostics_DiagnosticListener_IsEnabled_System_String_System_Object_System_Object_) which wouldn't be possible if we try to pre-cache things too much. Like, "I want to see events that involve resolving a certain type" to diagnose an issue would be a use case for being enabled based on certain data.

Reply:
Just pushed an update to have the event source `DiagnosticListener` check for `IsEnabled()` before checking if the specific named subscription is enabled. Benchmark the crap out if it, it should still basically boil down to a check for subscriptions != null before doing anything, which should be just as fast as what we had before. Granted, it's going to do that null check for each event rather than just one time, but that still should be pretty fast.